corehq/apps/accounting/tests/test_invoice_pdf.py

```diff
@@ -0,0 +1,171 @@
+        #
+        # TODO: if you add any of these, add new tests for them using
+        # `with self.assertRaises(ValueError):`.
+    ]
+
+    def iter_invoices(self):
+
+        def empty(*args, **kw):
+            """Callable used to prevent reportlab from building PDFs with
+            volatile values (ones that change on every render).
+            """
+            return b""
+
+        for invoice_id, kwargs in enumerate(self.iter_all_kwargs()):
+            invoice_name = f"invoice_{invoice_id:03}.pdf"
+            print(f"rendering invoice {invoice_name!r} with kwargs: {kwargs}")
+            invoice_path = os.path.join(self.TEST_DATA_DIR, invoice_name)
+            rendered = BytesIO()
+            # Patch these so the rendered PDF doesn't change even if the
+            # defaults in [local]settings.py change. Using `patch()` because
+            # `override_settings()` didn't work (cached values?).
+            with (
+                patch("corehq.apps.accounting.invoice_pdf.settings.BASE_ADDRESS", "server.local"),
+                patch("corehq.apps.accounting.invoice_pdf.settings.INVOICING_CONTACT_EMAIL", "bills@example.com"),
+            ):
+                # create a new invoice
+                invoice = self.get_invoice(rendered, **kwargs)
+
+                # add some line items
+                for item in range(4):
+                    invoice.add_item(
+                        f"Week {item + 1}",  # description
+                        "1",  # quantity
+                        Decimal("11.0"),  # unit_cost
+                        Decimal("11.0"),  # subtotal
+                        Decimal("-0.50"),  # credits
+                        Decimal("10.50"),  # total
+                    )
+
+                # Prevent /CreationDate and /ModDate metadata values
+                # see: https://stackoverflow.com/a/52359214
+                invoice.canvas.setDateFormatter(empty)
+
+                # Prevent /ID metadata value
+                # HACK: Canvas doesn't expose a setter to customize this
+                invoice.canvas._doc.ID = empty
+
+                # render the PDF
+                invoice.get_pdf()
+            yield invoice_path, rendered
+
+    def iter_all_kwargs(self):
+        """Yields every combination of kwargs possible from
+        ``INVOICE_KWARG_CHOICES`` excluding those in ``INVALID_INVOICE_KWARGS``.
+        """
+        # NOTE: depends on the fact that python dicts are ordered
+        keys = list(self.INVOICE_KWARG_CHOICES)
+        for argv in product(*self.INVOICE_KWARG_CHOICES.values()):
+            kwargs = {keys[i]: v for i, v in enumerate(argv)}
+            if kwargs not in self.INVALID_INVOICE_KWARGS:
+                yield kwargs
+
+    def get_invoice(self, file_or_path, *, is_wire, is_customer, is_prepayment):
+        """Returns an invoice instantiated with all available args passed by
+        position (positionally?).
```

Comment:
nit: pick one or the other? Both "passed positionally" or "passed by position" seem fine.

Reply:
I like "positionally", I'm going to change it.